KAFKA-7027: Add an overload build method in scala

massimosiani opened the pull request:
The Java API can pass a Properties object to StreamsBuilder#build, to allow, e.g., topology optimization, while the Scala API does not yet. The latter only delegates the work to the underlying Java implementation.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

massimosiani committed:
Add an overload build method in scala

massimosiani commented:
I didn' find a related issue, please let me know if I missed it (or missed a related PR), or if I should do anything else before submitting this.
Thank you!

bbejeck commented:
@massimosiani thanks for the contribution! This actually is related to KIP-312 (Jira https://issues.apache.org/jira/browse/KAFKA-7027) so I've updated the title.  

Also, can you rebase this to `trunk`?  Then we can cherry-pick to `2.2`

bbejeck reviewed (commented):
Thanks for the contribution @massimosiani! 
This PR looks good to me. 

We should add at least one test using the new builder overload.

bbejeck commented:
\cc any of (@guozhangwang, @mjsax, @vvcephei, and @ableegoldman) for a second review

massimosiani commented:
Hi @bbejeck, thanks for the review!
I have changed the base branch for the PR to trunk, and added a test case identical to the Simple one in TopologyTest.scala with the new API.
In the test, the TOPOLOGY_OPTIMIZATION property is passed, because that's the main goal of this PR.

Let me know if there is anything else I can do.
Thank you!

bbejeck reviewed (approved):
Thanks @massimosiani LGTM

bbejeck commented:
Java 8 failed with
```
org.apache.kafka.common.network.SslSelectorTest.testCloseOldestConnectionWithMultipleStagedReceives
kafka.api.UserQuotaTest.testThrottledProducerConsumer
kafka.network.SocketServerTest.testConnectionRateLimit
```
Java 11 passed
retest this please

vvcephei commented on streams/streams-scala/src/test/scala/org/apache/kafka/streams/scala/TopologyTest.scala:
```diff
@@ -268,4 +269,41 @@ class TopologyTest extends JUnitSuite {
+      val streamBuilder = new StreamsBuilderJ
+      val textLines = streamBuilder.stream[String, String](inputTopic)
+
+      val _: KStreamJ[String, String] = textLines.flatMapValues(
```
Just curious, what's the advantage of assigning to `_` instead of just ignoring the result of the builder calls?

massimosiani replied:
Hi, @vvcephei, there are other tests in that class, and I wanted to preserve the general structure. Probably, they should be clean up. Removed from my code.

vvcephei commented on streams/streams-scala/src/test/scala/org/apache/kafka/streams/scala/TopologyTest.scala:
```diff
@@ -268,4 +269,41 @@ class TopologyTest extends JUnitSuite {
+    }
+
+    // should match
+    assertEquals(getTopologyScala, getTopologyJava)
```
This comparison actually is invalid. No reason you should know that, though...

A contributor recently pointed it out (see https://github.com/apache/kafka/pull/6210).

As a workaround, can we try comparing the result of `TopologyDescription#toString` instead?

massimosiani replied:
With the new test, this command works:
gradle streams:streams-scala:test --tests TopologyTest
(gradle 5.2.1 on MacOS)

bbejeck replied:
@massimosiani @vvcephei is correct about the test

> This comparison actually is invalid. No reason you should know that, though...
>A contributor recently pointed it out (see #6210).
>As a workaround, can we try comparing the result of TopologyDescription#toString instead?

~~When comparing the `getTopologyScala.toString` and `getTopologyJava.toString`  the test fails.  But it's not your fault though, I'm sure you have just exposed a bug! Thanks and I've created a Jira for it (https://issues.apache.org/jira/browse/KAFKA-8101) and when that is fixed we can get this merged.~~

~~EDIT: This looks like a bug with the Streams Scala API in general as when using `StreamsConfig.NO_OPTIMIZATION` or not passing in `Properties` to `builder.build()` the  test still fails.~~

vvcephei commented on streams/streams-scala/src/test/scala/org/apache/kafka/streams/scala/TopologyTest.scala:
```diff
@@ -268,4 +269,41 @@ class TopologyTest extends JUnitSuite {
+      val _: KStream[String, String] =
+        textLines.flatMapValues(v => pattern.split(v.toLowerCase))
+
+      streamBuilder.build(props).describe()
```
It seems like, if the intent is to make sure the optimization property actually gets applied, we'd need to use a topology that would change under optimization. Perhaps you can copy the topology (and verification logic) from `org.apache.kafka.streams.integration.RepartitionOptimizingIntegrationTest`?

Alternatively, it might be sufficient to make it a unit test by injecting a mocked Java builder and just verify that the Scala layer passes the properties down correctly... Not sure how straightforward that would be, though.

massimosiani replied:
It's been a pain to write the "Java" version (that's probably because of me being new to kafka), but I've done it.

vvcephei reviewed (changes requested):
Hi @massimosiani , Thanks for the contribution!

The implementation looks "obviously correct", but I had some comments on the test. The biggest concern is that the equality comparison of `TopologyDescription` is currently broken.

As I said, the implementation looks right, but the test would have value as a regression test, to make sure we don't break the logic in the future.

Thanks again,
-John

bbejeck commented:
@massimosiani After looking over the Scala API the issue for the different topology between Java and Scala is here https://github.com/apache/kafka/blob/trunk/streams/streams-scala/src/main/scala/org/apache/kafka/streams/scala/kstream/KGroupedStream.scala#L51 .

`KGroupedStream#count` in Scala needs to add an additional `mapValues` proceessor to convert from a Java Long to a Scala Long, so a `count` operation won't have the same topology.  

Thanks for your hard work so far and here's what I'd suggest going forward.  

Change the example to have 3 streams and you call either `selectKey` or `map` on the first stream instance.  Then perform two joins, 

```scala
val sourceStream = builder.stream(inputTopic, Consumed.`with`(Serdes.String, Serdes.String))
val mappedStream: KStreamJ[String, String] =sourceStream.map((k: String, v: String) => ...))

val stream2 = ...
val stream3 = ...

val join1 = mappedStream.join(stream2....) 
val join2 - mappedStream.join(stream3,...)
```

This way with optimization turned on we should end up with only one repartition topic and since there is no aggregation operation the toplogies should end up the same.

We'll stil want the assertion step to be `assertEquals(getTopologyScala.toString, getTopologyJava.toString)`

massimosiani committed:
KAFKA-7027 change the test to exclude counts

massimosiani commented:
Hi @bbejeck, thank you.
I am not sure whether it makes sense, but I also added an assertion to check that the topology without optimization is different from the optimized one, so I am actually testing an optimized topology.
I had to add the explicit NO_OPTIMIZE property so we are not checking the default behavior of build().
Let me know if that makes sense at all.
Thank you!

bbejeck commented on streams/streams-scala/src/test/scala/org/apache/kafka/streams/scala/TopologyTest.scala:
```diff
@@ -268,4 +276,134 @@ class TopologyTest extends JUnitSuite {
+      builder
+    }
+
+    assertNotEquals(getTopologyScala.build(props).describe.toString,
```
I like the idea here, but I'm thinking it's a good idea to compare unoptimized Scala topology to the unoptimized Java topology so I'd change this line to 
```scala
assertEquals(getTopologyScala.build(propsNoOptimization).describe.toString,
                    getTopologyScala.build(propsNoOptimization).describe.toString)
```

massimosiani replied:
What about adding your assertion instead of substituting to line 399? Do you agree?

bbejeck replied:
>What about adding your assertion instead of substituting to line 399? 

Yes, that works as well

bbejeck reviewed (commented):
@massimosiani thanks for the updates.

>I am not sure whether it makes sense, but I also added an assertion to check that the topology without optimization is different from the optimized one, so I am actually testing an optimized topology.

Yes, that makes sense, but I think to compare both optimized and unoptimized Scala/Java topologies has merit, I've left a comment regarding that.

Also, the build failure is related.  The code you have here works for Scala version `2.12` but Scala `2.11` doesn't support Java 8 lamdas so the `Java 8 Scala 2.11` build is failing.

You'll need to change `initializer` , `aggregator` etc. to have Java anonymous classes 

```scala
val initializer: Initializer[Integer] =  new Initializer[Integer] {
     override def apply(): Integer = 0
 }
 val aggregator: Aggregator[String, String, Integer] = new Aggregator[String, String, Integer] {
     override def apply(key: String, value: String, aggregate: Integer): Integer =  aggregate + value.length
 }
   ```
This applies to all the operators on lines 345 to 353 I didn't list them out here
If you run `./gradlew -PscalaVersion=2.11 :streams:streams-scala:compileTestScala` you'll see what all the errors are.

Thanks for your hard work and patience on this!

massimosiani committed:
fix for scala 2.11

bbejeck reviewed (approved):
Thanks for the updates @massimosiani.  Changes LGTM pending Jenkins passing

bbejeck commented:
Moving to new comment to make sure this gets picked up

@massimosiani one last thing I just noticed, do you mind changing the `assertEquals` on line 432-433
from:

`assertEquals(getTopologyScala.build(propsNoOptimization).describe.toString,
                 getTopologyScala.build(propsNoOptimization).describe.toString)`
to:

`assertEquals(getTopologyScala.build(propsNoOptimization).describe.toString, getTopologyJava.build(propsNoOptimization).describe.toString)`

My apologies as this is my error from my previous comment.

massimosiani committed:
fix an assertion

massimosiani commented:
> Moving to new comment to make sure this gets picked up
> 
> @massimosiani one last thing I just noticed, do you mind changing the `assertEquals` on line 432-433
> from:
> 
> `assertEquals(getTopologyScala.build(propsNoOptimization).describe.toString, getTopologyScala.build(propsNoOptimization).describe.toString)`
> to:
> 
> `assertEquals(getTopologyScala.build(propsNoOptimization).describe.toString, getTopologyJava.build(propsNoOptimization).describe.toString)`
> 
> My apologies as this is my error from my previous comment.

My fault, I took it for granted and did not read it carefully.

vvcephei reviewed (approved):
LGTM! Thanks, @massimosiani !

bbejeck commented:
Thanks for the contribution @massimosiani!

Merged #6373 to trunk

bbejeck commented:
cherry-picked to 2.2, and 2.1
